### PR TITLE
SE-0070: Make Optional Requirements Objective-C-only.

### PR DESCRIPTION
**Pull request by koke**
#### What's in this pull request?

Optional protocol requirements now require an explicit @objc attribute.
#### Resolved bug number: ([SR-1395](https://bugs.swift.org/browse/SR-1395))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by tkremenek** [comment]
This looks great to me.  Waiting for testing to finish.

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -7808,9 +7808,9 @@ static void validateAttributes(TypeChecker &TC, Decl *D) {
       TC.diagnose(OA->getLocation(),
                   diag::optional_attribute_non_protocol);
       D->getAttrs().removeAttribute(OA);
-    } else if (!cast<ProtocolDecl>(D->getDeclContext())->isObjC()) {
+    } else if (!D->getAttrs().getAttribute<ObjCAttr>()) {
       TC.diagnose(OA->getLocation(),
-                  diag::optional_attribute_non_objc_protocol);
+                  diag::optional_attribute_non_objc_attribute);
       D->getAttrs().removeAttribute(OA);
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: The recovery here should be better: there is no reason to drop the `optional` when it's just missing the `@objc`. I suggest leaving the original error in place. Then, add a separate check that complains about a missing, explicit '@objc' and provides a Fix-It, e.g.:

    if (!D->getAttrs().getAttribute<ObjCAttr>()) {
        TC.diagnose(OA->getLocation(), diag::optional_attribute_non_objc_attribute);
          .fixItInsert(D->getAttributeInsertionLoc(false), "@objc ")
    }

File: test/decl/protocol/req/optional.swift [hunk 2]
```diff
@@ -219,10 +219,14 @@ optional class optErrorClass { // expected-error{{'optional' modifier cannot be
 }
   
 protocol optErrorProtocol {
-  optional func foo(_ x: Int) // expected-error{{'optional' can only be applied to members of an @objc protocol}}
+  optional func foo(_ x: Int) // expected-error{{'optional' requirements are an Objective-C compatibility feature; add '@objc'}}
```
Comment (DougGregor) [inline test/decl/protocol/req/optional.swift]: Per my above comment, the original error is correct here.

File: test/decl/protocol/req/optional.swift [hunk 3]
```diff
@@ -219,10 +219,14 @@ optional class optErrorClass { // expected-error{{'optional' modifier cannot be
...
+}
+
+@objc protocol optObjcAttributeProtocol {
+  optional func foo(_ x: Int) // expected-error{{'optional' requirements are an Objective-C compatibility feature; add '@objc'}}
```
Comment (DougGregor) [inline test/decl/protocol/req/optional.swift]: To test the Fix-It I mentioned about, add {{2-2=@objc }} following the expected-error's }}.

**Comment by DougGregor** [comment]
Thanks for working on this! It's almost there.

**Comment by DougGregor** [comment]
I'm going to accept this PR as-is and apply my corrections in a follow-up commit.

**Comment by DougGregor** [comment]
My post-merge tweaks are in https://github.com/apple/swift/commit/bc66b1961c739811f508b96a62f8425ea115c54b

**Comment by koke** [comment]
Thanks for taking care of the Fix-It 😄

**Pull request merged**